lib/srv/app/aws/assume_role.go

```diff
@@ -0,0 +1,114 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aws
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+
+	"github.com/gravitational/teleport/lib/srv/app/common"
+	"github.com/gravitational/teleport/lib/tlsca"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+func updateAssumeRoleDuration(identity *tlsca.Identity, w http.ResponseWriter, req *http.Request, clock clockwork.Clock) error {
+	// Skip non-AssumeRole request
+	query, found, err := getAssumeRoleQuery(req)
+	if err != nil || !found {
+		return trace.Wrap(err)
+	}
+
+	// Deny access if identity duration is shorter than the minimum that can be
+	// requested.
+	identityTTL := identity.Expires.Sub(clock.Now())
+	if identityTTL < assumeRoleMinDuration {
+		// TODO write error message in XML so the client can understand.
+		return trace.AccessDenied("minimum AWS session duration is %v but Teleport identity expires in %v", assumeRoleMinDuration, identityTTL)
+	}
+
+	// Use shorter requested duration (no update required).
+	if getAssumeRoleQueryDuration(query) <= identityTTL {
+		return nil
+	}
+
+	// Rewrite the request.
+	if err := rewriteAssumeRoleQuery(req, withAssumeRoleQueryDuration(query, identityTTL)); err != nil {
+		return trace.Wrap(err)
+	}
+	w.Header().Add(common.TeleportAPIInfoHeader, fmt.Sprintf("requested DurationSeconds of AssumeRole is overwritten to \"%d\" as the Teleport identity will expire at %v", int(identityTTL.Seconds()), identity.Expires))
```

Comment:
Nit: make it clear we are shortening the duration.

```suggestion
	w.Header().Add(common.TeleportAPIInfoHeader, fmt.Sprintf("requested DurationSeconds of AssumeRole is lowered to \"%d\" as the Teleport identity will expire at %v", int(identityTTL.Seconds()), identity.Expires))
```